### PR TITLE
Fix classloading issue with OutputNode PlanNode

### PR DESCRIPTION
**Pull request by jainxrohit**
OutputNode plan class is used for creating logical plan. Now logical plan creation is part of the pluggable analyzer interface, respective analyzer plugins are expected to use OutputNode class.

Previously, the OutputNode class was extended from InternalPlanNode, which required it to be available with AppClassLoader. However, since the analyzer plugins now instantiate the OutputNode class, it and its dependencies are loaded by PluginClassLoader, making it unavailable for the rest of the AppClassLoader.

To fix this issue, we are extending OutputNode from PlanNode instead of InternalPlanNode and also moving it to the spi package, as it is anyways expected to be leveraged by plugins. This also whitelists this class for PluginClassLoader.

The other alternative would have been to add com.facebook.presto.sql.planner package to the whitelists for PluginClassLoader, however that would have added to many classes to the whitelist which is not necessary.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-spi/src/main/java/com/facebook/presto/spi/plan/OutputNode.java [hunk 1]
```diff
@@ -61,17 +58,19 @@ public OutputNode(
...
         requireNonNull(columnNames, "columnNames is null");
-        Preconditions.checkArgument(columnNames.size() == outputVariables.size(), "columnNames and assignments sizes don't match");
+        if (columnNames.size() != outputVariables.size()) {
+            throw new IllegalArgumentException("columnNames and assignments sizes don't match");
+        }
 
         this.source = source;
         this.columnNames = columnNames;
-        this.outputVariables = ImmutableList.copyOf(outputVariables);
+        this.outputVariables = unmodifiableList(outputVariables);
```
Comment (mlyublena) [inline presto-spi/src/main/java/com/facebook/presto/spi/plan/OutputNode.java:67]: why did this change from Immutable to unmodifiable?

**Comment by jainxrohit** [comment]
> why is this specific to output node and not all plan nodes?

For all plan nodes that we anticipate a pluggable analyzer to use, this would hold true. However, for certain internal plan nodes (such as ExchangeNode), which are not utilized by the analyzer, this may not be the case. In the present range of query shapes that Crux supports, the only missing component was the OutputNode.

**Review by mlyublena — APPROVED** [review]

**Comment by rschlussel** [comment]
> > why is this specific to output node and not all plan nodes?
> 
> For all plan nodes that we anticipate a pluggable analyzer to use, this would hold true. However, for internal plan nodes (such as ExchangeNode), which are not utilized by the analyzer, this would not be the case. In the present range of query shapes that Crux supports, the only missing class was the OutputNode.

That would include most plan nodes then. Currently very few plan nodes are in the spi. we'll need to add almost all of them over.

**Comment by jainxrohit** [comment]
> 

Not all, only those which are required while constructing logical plan nodes. Nodes added during optimization phase are not required. 
So either we can move these one by one as we construct shapes in the CRUX (which I feel is cleaner approach, as classes used by plugins should be in SPI),  or we can add current main class planner package to whitelist.

**Comment by jainxrohit** [comment]
> facebook

I discussed offline with @rschlussel and we have agreed to go with approach 2. I will update the PR by fixing more plan nodes, which would be used by analyzers plugins.

**Review by elharo — COMMENTED** [review]
com.facebook.presto.sql.planner.plan.OutputNode appears to be public API. Who's affected by removing it?

File: presto-spi/src/main/java/com/facebook/presto/spi/plan/OutputNode.java (hunk 1 above)
Comment (elharo) [reply presto-spi/src/main/java/com/facebook/presto/spi/plan/OutputNode.java:67]: It's also a good idea to replace Guava and similar libraries with JDK methods where available. A lot of Guava, Apache commons collections, and the like exist primarily to provide utilities that are available in JDK 8+ for earlier JDK versions.

**Comment by jainxrohit** [comment]
> com.facebook.presto.sql.planner.plan.OutputNode appears to be public API. Who's affected by removing it?

@elharo we are not removing it, we have moved to spi package, which makes it even more official. SPI package in general is very well understood by plugins owner.

**Review by rschlussel — APPROVED** [review]
We discussed, and not all nodes can be easily moved to the spi, so some work is needed there.  We'll start with just the OutputNode to unblock testing an end to end flow for a simple query with a plugin analyzer.  We'll migrate the rest of the nodes afterwards.

**Comment by mbasmanova** [comment]
@jainxrohit Rohit, it seem like this change broken Sapphire-Native. CC: @miaoever @tanjialiang

**Comment by amitkdutta** [comment]
@jainxrohit This is a protocol change and it broke native worker. e2e tests failed in this PR as well. It would require generate native protocol bindings, and make e2e tests pass.

https://github.com/prestodb/presto/actions/runs/4547330179/jobs/8017077942?pr=19265

 com.facebook.presto.spi.PrestoException: Expected response code from http://127.0.0.1:1237/v1/task/20230328_223459_00023_3bbj6.0.0.0?summarize to be 200, but was 500: .OutputNode no abstract type PlanNode

**Comment by jainxrohit** [comment]
> We discussed, and not all nodes can be easily moved to the spi, so some work is needed there. We'll start with just the OutputNode to unblock testing an end to end flow for a simple query with a plugin analyzer. We'll migrate the rest of the nodes afterwards.

Fixed in the https://github.com/prestodb/presto/pull/19284

**Comment by jainxrohit** [comment]
> @jainxrohit This is a protocol change and it broke native worker. e2e tests failed in this PR as well. It would require generate native protocol bindings, and make e2e tests pass.
> 
> https://github.com/prestodb/presto/actions/runs/4547330179/jobs/8017077942?pr=19265
> 
> com.facebook.presto.spi.PrestoException: Expected response code from http://127.0.0.1:1237/v1/task/20230328_223459_00023_3bbj6.0.0.0?summarize to be 200, but was 500: .OutputNode no abstract type PlanNode

Fixed in the https://github.com/prestodb/presto/pull/19284

**Comment by jainxrohit** [comment]
> @jainxrohit Rohit, it seem like this change broken Sapphire-Native. CC: @miaoever @tanjialiang

Fixed in the https://github.com/prestodb/presto/pull/19284

**Pull request merged**